File: blurb/app/src/main/java/com/mowdowndevelopments/blurb/ui/navHost/MainViewModel.java
package com.mowdowndevelopments.blurb.ui.navHost;

import android.app.Application;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.mowdowndevelopments.blurb.R;
import com.mowdowndevelopments.blurb.network.LoadingStatus;
import com.mowdowndevelopments.blurb.network.ResponseModels.AutoCompleteResponse;
import com.mowdowndevelopments.blurb.network.Singletons;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import timber.log.Timber;

public class MainViewModel extends AndroidViewModel {


    private MutableLiveData<List<AutoCompleteResponse>> autoCompleteDialogData;
    private MutableLiveData<LoadingStatus> logoutStatus;
    private MutableLiveData<String> errorMessage;
    private boolean loadingForDialog = false;

    public MainViewModel(@NonNull Application application) {
        super(application);
        errorMessage = new MutableLiveData<>();
        logoutStatus = new MutableLiveData<>(LoadingStatus.WAITING);
        autoCompleteDialogData = new MutableLiveData<>();
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public LiveData<LoadingStatus> getLogoutStatus() {
        return logoutStatus;
    }

    public LiveData<List<AutoCompleteResponse>> getAutoCompleteDialogData() {
        return autoCompleteDialogData;
    }

    public void clearAutoCompleteDialogData(){
        autoCompleteDialogData.setValue(null);
    }

    public void logout(){
        logoutStatus.postValue(LoadingStatus.LOADING);
        Singletons.getNewsBlurAPI(getApplication()).logout().enqueue(new Callback<Void>() {
            @Override
            public void onResponse(@NotNull Call<Void> call, @NotNull Response<Void> response) {
                if (response.isSuccessful()){
                    logoutStatus.postValue(LoadingStatus.DONE);
                    SharedPreferences prefs = getApplication()
                            .getSharedPreferences(getApplication().getString(R.string.shared_pref_file), 0);
                    prefs.edit().putBoolean(getApplication().getString(R.string.logged_in_key), false).apply();
                } else {
                    logoutStatus.postValue(LoadingStatus.ERROR);
                    //TODO Add messages to explain HTTP Error codes in plain language
                    String errorMsg = getApplication().getString(R.string.http_error, response.code());
                    errorMessage.postValue(errorMsg);
                    Timber.e("onResponse: %s", errorMsg);
                }
            }

            @Override
            public void onFailure(@NotNull Call<Void> call, @NotNull Throwable t) {
                logoutStatus.postValue(LoadingStatus.ERROR);
                Timber.e(t, "onFailure: %s", t.getMessage());
                errorMessage.postValue(t.getLocalizedMessage());
                FirebaseCrashlytics.getInstance().log(String.format("loginCallback.onFailure: %s", t.getMessage()));
                FirebaseCrashlytics.getInstance().recordException(t);
            }
        });
    }

    //Used for NewFeedDialogFragment;
    public void loadDataForFeedAutoComplete(String searchTerm){
        if (loadingForDialog) return;
        loadingForDialog = true;
        Singletons.getNewsBlurAPI(getApplication()).getAutoCompleteResults(searchTerm).enqueue(new Callback<List<AutoCompleteResponse>>() {
            //Will fail silently
            @Override
            public void onResponse(@NotNull Call<List<AutoCompleteResponse>> call, @NotNull Response<List<AutoCompleteResponse>> response) {
                loadingForDialog = false;
                if (response.isSuccessful()){
                    autoCompleteDialogData.postValue(response.body());
                }
            }

            @Override
            public void onFailure(@NotNull Call<List<AutoCompleteResponse>> call, @NotNull Throwable t) {
                loadingForDialog = false;
            }
        });
    }
}

File: blurb/app/src/main/java/com/mowdowndevelopments/blurb/ui/navHost/MainViewModel.kt
package com.mowdowndevelopments.blurb.ui.navHost

import android.app.Application
import androidx.core.content.edit
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.mowdowndevelopments.blurb.R
import com.mowdowndevelopments.blurb.network.LoadingStatus
import com.mowdowndevelopments.blurb.network.Singletons.getNewsBlurAPI
import com.mowdowndevelopments.blurb.network.responseModels.AutoCompleteResponse
import kotlinx.coroutines.launch
import timber.log.Timber

class MainViewModel(application: Application) : AndroidViewModel(application) {
    private val _autoCompleteDialogData: MutableLiveData<List<AutoCompleteResponse>?> = MutableLiveData()
    val autoCompleteDialogData: LiveData<List<AutoCompleteResponse>?>
        get() = _autoCompleteDialogData
    private val _logoutStatus: MutableLiveData<LoadingStatus> = MutableLiveData(LoadingStatus.WAITING)
    val logoutStatus: LiveData<LoadingStatus>
        get() = _logoutStatus
    private val _inAppDialogStatus: MutableLiveData<LoadingStatus> = MutableLiveData(LoadingStatus.WAITING)
    val inAppDialogStatus: LiveData<LoadingStatus>
        get() = _inAppDialogStatus
    private val _errorMessage: MutableLiveData<String> = MutableLiveData()
    val errorMessage: LiveData<String>
        get() = _errorMessage

    private var loadingForDialog = false
    private var purchaseRetryAttempts = 0

    fun resetRetryAttempts() {
        purchaseRetryAttempts = 0
    }

    fun incrementRetryAttempts() {
        ++purchaseRetryAttempts
    }

    fun canKeepRetryingPurchase(): Boolean = purchaseRetryAttempts < MAX_RETRY_ATTEMPTS

    fun postNewErrorMessage(message: String) = _errorMessage.postValue(message)

    fun postInAppDialogStatus(status: LoadingStatus) = _inAppDialogStatus.postValue(status)

    fun clearAutoCompleteDialogData() {
        _autoCompleteDialogData.value = null
    }

    fun logout() {
        _logoutStatus.postValue(LoadingStatus.LOADING)
        viewModelScope.launch {
            try {
                val c = getApplication<Application>().applicationContext
                val response = getNewsBlurAPI(c).logout()
                if (response.isSuccessful) {
                    _logoutStatus.postValue(LoadingStatus.DONE)
                    c.getSharedPreferences(c.getString(R.string.shared_pref_file), 0).edit {
                        putBoolean(c.getString(R.string.logged_in_key), false) }
                } else {
                    _logoutStatus.postValue(LoadingStatus.ERROR)
                    //TODO Add messages to explain HTTP Error codes in plain language
                    val errorMsg = c.getString(R.string.http_error, response.code())
                    _errorMessage.postValue(errorMsg)
                    Timber.e("onResponse: %s", errorMsg)
                }
            } catch (e: Exception) {
                _logoutStatus.postValue(LoadingStatus.ERROR)
                Timber.e(e, "onFailure: %s", e.message)
                _errorMessage.postValue(e.localizedMessage)
                FirebaseCrashlytics.getInstance().log(String.format("loginCallback.onFailure: %s", e.message))
                FirebaseCrashlytics.getInstance().recordException(e)
            }
        }
    }

    //Used for NewFeedDialogFragment;
    fun loadDataForFeedAutoComplete(searchTerm: String?) {
        if (loadingForDialog) return
        loadingForDialog = true
        viewModelScope.launch {
            try {
                val response = getNewsBlurAPI(getApplication())
                        .getAutoCompleteResults(requireNotNull(searchTerm))
                if (response.isSuccessful) {
                    _autoCompleteDialogData.postValue(response.body())
                }
            } catch (e: Exception) { //Fail silently
            } finally {
                loadingForDialog = false
            }
        }
    }

    companion object {
        private const val MAX_RETRY_ATTEMPTS = 10
    }

}